Add a Webhook

jhuleatt opened the pull request:
Adds a new webhook to subscribe to Invoice lifecycle events. When an event is received, it updates the status of the invoice in Firestore. This way, a developer can subscribe to the invoice document in Firestore and get realtime updates as it changes.

Also changes the firestore-triggered function - now it writes the Invoice ID into the Firestore document.

jhuleatt commented on functions/src/index.ts:
```diff
@@ -73,9 +73,9 @@ export const sendInvoice = functions.handler.firestore.document.onCreate(
-      let customers: Stripe.ApiList<
-        Stripe.Customer
-      > = await stripe.customers.list({ email: payload.email });
+      let customers: Stripe.ApiList<Stripe.Customer> = await stripe.customers.list(
```
this is Prettier noise

jhuleatt commented on functions/src/index.ts:
```diff
@@ -101,6 +101,12 @@ export const sendInvoice = functions.handler.firestore.document.onCreate(
+        // so that we can find it in the webhook
+        await snap.ref.update({
+          stripeInvoiceId: invoice.id
+        });
```
FYI I added this so that we can look up these documents in the webhook

jhuleatt commented on functions/src/index.ts:
```diff
@@ -120,3 +126,56 @@ export const sendInvoice = functions.handler.firestore.document.onCreate(
+  "invoice.finalized",
+  "invoice.payment_failed",
+  "invoice.payment_succeeded"
+]);
```
Are these the events we should care about? Should we take any out, or are there some other useful ones to add?

adreyfus-stripe replied:
I would also add these, they could happen via actions from the Stripe Dashboard or other API calls once the Invoice is opened:

```
invoice.payment_action_required
invoice.voided
invoice.marked_uncollectible
```

jhuleatt replied:
added, thanks!

jhuleatt commented on functions/src/index.ts:
```diff
@@ -120,3 +126,56 @@ export const sendInvoice = functions.handler.firestore.document.onCreate(
+
+    const doc = invoicesInFirestore.docs[0];
+    await doc.ref.update({
+      stripeInvoiceStatus: event
```
Is the event the useful thing to store in Firestore? Or should it be something else from the event object?

adreyfus-stripe replied:
Hard to know I'm not super confident what bits are useful for developers! One idea:

```
const invoiceStatus = event === 'invoice.payment_failed' ? 'payment_failed' : invoice.status

invoiceStatus,
invoiceRecord: `https://dashboard.stripe.com/invoices/${invoice.id}
```

That can keep track of the status (one of draft, open, paid, uncollectible, or void) with a special status for payment_failed because I think that would still keep the invoice `open`? So some kind of indicator that someone tried to pay might be useful. 

A lot of this extra can be found in the Stripe dashboard, so I included a link to the record in the Stripe dashboard in case the developer wants to dig deeper.

adreyfus-stripe replied:
Actually, thinking about my proposal we can drop the link to the invoice in the Stripe Dashboard and just record the invoiceStatus for now (if we want to include a link we can do so in the main Firebase function, it doesn't have to be in the webhook since it will never change)

jhuleatt replied:
Thanks for the suggestion! I changed the fields written to Firestore to be:

- `stripeInvoiceStatus`: The field you suggested
- `lastStripeEvent`: The event name that last triggered the webhook for this invoice

Do those sound good?

I'll add the link to the dashboard in the other function since this PR already modifies that function to write the invoice id

jhuleatt commented on package.json:
```diff
@@ -22,5 +22,9 @@
+  "prettier": {
+    "singleQuote": false,
+    "printWidth": 80
+  },
```
I added this to cut down on Prettier noise (I think my default Prettier config is different from yours so I was getting really noisy diffs)

jhuleatt commented on test-params.env:
```diff
@@ -0,0 +1,4 @@
+LOCATION=us-central1
+DB_PATH=invoices
+STRIPE_SECRET_KEY=sk_test_CtBQPkkX3am6OFnZuVXmbs0E00LfmkG48g
```
since this is a test key, is it safe to put in github? or should I take it down?

(this file will allow for easier testing)

adreyfus-stripe replied:
Ah yeah, we should take it down. We discourage our developers from committing any keys in GitHub, even test keys, because someone could get the wrong idea.

jhuleatt replied:
took it down

adreyfus-stripe commented on functions/lib/index.js:
```diff
@@ -104,4 +109,80 @@ exports.sendInvoice = functions.handler.firestore.document.onCreate(async (snap)
     }
     return;
 });
+// call this function to easily
```
Is this something we want to keep? I like it as an easy way to seed some test data, curious if other extensions do similar things.

adreyfus-stripe replied:
Ah just noticed that this isn't in the typescript code below

jhuleatt replied:
I rebuilt, this should be gone now.

adreyfus-stripe commented on functions/src/index.ts:
```diff
@@ -120,3 +126,56 @@ export const sendInvoice = functions.handler.firestore.document.onCreate(
+export const updateInvoice = functions.handler.https.onRequest(
+  async (req, resp) => {
+    let invoice: Stripe.Invoice;
+    let event;
```
We have a type for that!

```
let event: Stripe.Event;
```

jhuleatt replied:
added, thanks!

adreyfus-stripe commented on functions/src/index.ts:
```diff
@@ -120,3 +126,56 @@ export const sendInvoice = functions.handler.firestore.document.onCreate(
+    let event;
+
+    try {
+      invoice = req.body.data.object as Stripe.Invoice;
```
I think it would be good to check the webhook signature -- this guarantees the request comes from Stripe 

```
    try {
      event = stripe.webhooks.constructEvent(
        req.body,
        req.headers["stripe-signature"],
        process.env.STRIPE_WEBHOOK_SECRET
      );
    } catch (err) {
      console.log(`⚠️  Webhook signature verification failed.`);
      res.sendStatus(400);
      return;
    }
```

We need to pass the raw request body to Stripe to verify the signature, is there any way to access that from the function? With Express I was able to add something like bodyParser.raw({ type: "application/json" }) to retrieve the buffer

jhuleatt replied:
Ohh, I was wondering why the [example code](https://stripe.com/docs/webhooks/build#example-code) used `bodyParser.raw()`. I think Cloud Functions exposes `req.rawBody`. I'll give it a shot

jhuleatt replied:
added!

adreyfus-stripe commented:
Thank you for this! I just tried running it and ran into some permissions issues: 

1. updateInvoice was not publicly accessible and throwing this error:
``` 
<html><head><meta http-equiv="content-type" content="text/html;charset=utf-8"><title>403 Forbidden</title></head><body text=#000000 bgcolor=#ffffff><h1>Error: Forbidden</h1><h2>Your client does not have permission to get URL <code>/ext-create-stripe-invoice-updateInvoice</code> from this server.</h2><h2></h2></body></html>
```
I fixed this by making the function accessible to the public by setting `allUsers`. Is there something in extension.yaml we could add to set that up so it's reachable by Stripe? 

2. Could not update the Firestore documents, was getting:
```
[Error] Unhandled error occurred during processing: { Error: 7 PERMISSION_DENIED: Missing or insufficient permissions. at Object.callErrorFromStatus (/srv/node_modules/@grpc/grpc-js/build/src/call.js:30:26) at Http2CallStream.call.on (/srv/node_modules/@grpc/grpc-js/build/src/client.js:96:33) at emitOne (events.js:121:20) at Http2CallStream.emit (events.js:211:7) at process.nextTick (/srv/node_modules/@grpc/grpc-js/build/src/call-stream.js:100:22) at _combinedTickCallback (internal/process/next_tick.js:132:7) at process._tickDomainCallback (internal/process/next_tick.js:219:9) code: 7, details: 'Missing or insufficient permissions.', metadata: Metadata { internalRepr: Map {}, options: {} }, note: 'Exception occurred in retry method that was not classified as transient' }
```

I couldn't actually figure this one out. I played around with the security rules on my Firestore db and allowed read, write if true for all documents. 

Once we clear this up I think we can merge

jhuleatt commented:
Thanks for trying it out and catching those issues!

1. I've now added instructions to the postinstall in our shared doc to tell installers how to make the function publicly callable. Unfortunately, we can't do anything in extension.yaml to automatically configure.
2. I've updated extension.yaml to declare the correct permissions so that the extension can write to Firestore.

I installed it on a personal project and tried the whole workflow, and it worked!

adreyfus-stripe commented:
Great, I also just verified the workflow!
